libvast/src/segment_store.cpp

```diff
@@ -157,43 +157,173 @@ std::unique_ptr<store::lookup> segment_store::extract(const ids& xs) const {
+    // Get all slices in the segment and generating a new segment that contains
+    // only what's left after dropping the selection.
+    auto segment_ids = sref.meta().get_flat_slice_ids();
+    std::vector<table_slice_ptr> slices;
+    if (auto maybe_slices = sref.lookup(segment_ids)) {
+      using std::swap;
+      swap(slices, *maybe_slices);
+      if (slices.empty()) {
+        VAST_WARNING(this, "got no slices after lookup for segment", segment_id,
+                     "=> erases entire segment!");
+        drop(sref);
+        return;
+      }
+    } else {
+      VAST_WARNING(this, "was unable to get table slice for segment",
+                   segment_id, "=> erases entire segment!");
+      drop(sref);
+      return;
+    }
+    VAST_ASSERT(slices.size() > 0);
+    auto keep_mask = ~xs;
+    std::vector<table_slice_ptr> new_slices;
+    for (auto& slice : slices) {
+      // Expand keep_mask on-the-fly if needed.
+      auto max_id = slice->offset() + slice->rows();
+      if (keep_mask.size() < max_id)
+        keep_mask.append_bits(true, max_id - keep_mask.size());
+      size_t new_slices_size_before = new_slices.size();
+      select(new_slices, slice, keep_mask);
+      size_t remaining_rows = 0;
+      for (size_t i = new_slices_size_before; i < new_slices.size(); ++i)
+        remaining_rows += new_slices[i]->rows();
+      erased_events += slice->rows() - remaining_rows;
+    }
+    if (new_slices.empty()) {
+      VAST_WARNING(this, "was unable to generate any new slice for segment",
+                   segment_id, "=> erases entire segment!");
+      drop(sref);
+      return;
+    }
+    VAST_DEBUG(this, "shrinks segment", segment_id, "from", slices.size(), "to",
+               new_slices.size(), "slices");
+    // Remove stale state.
+    drop_range(sref.id());
+    // Create a new segment from the remaining slices.
+    segment_builder tmp_builder;
+    segment_builder* builder = &tmp_builder;
+    if constexpr (std::is_same_v<decltype(sref), segment_builder&>) {
+      sref.reset();
+      builder = &sref;
+    }
+    for (auto& slice : new_slices) {
+      if (auto err = builder->add(slice)) {
+        VAST_ERROR(this, "failed to add slice to builder:" << err);
+      } else if (!segments_.inject(slice->offset(),
+                                   slice->offset() + slice->rows(),
+                                   builder->id()))
+        VAST_ERROR(this, "failed to update range_map");
+    }
+    // Flush the new segment and remove the previous segment.
+    if constexpr (std::is_same_v<decltype(sref), segment&>) {
+      auto new_segment = builder->finish();
+      auto filename = segment_path() / to_string(new_segment->id());
+      if (auto err = save(nullptr, filename, new_segment))
+        VAST_ERROR(this, "failed to persist the new segment");
+      if (auto err = save(nullptr, meta_path(), segments_))
+        VAST_ERROR(this, "failed to persist meta data after adding segment");
```

Comment:
Could this be done maybe at the very end of the function as opposed to in the loop?

Reply:
Absolutely.